File: pmid.ts
export const PMIDPattern: string = "^[0-9]+$";
export const rePMID: RegExp = new RegExp(PMIDPattern);

/**
 * normalizePMID normalized the Pub Med ID.
 *
 * @param pubMedID, a string representing a pmid.
 * @returns string
 */
export function normalizePMID(pubMedID: string): string {
  return pubMedID.replace(/pmid:\s+|pmid:/i, "").trim();
}

/**
 * validatePMID will check a if a pmid is valid after normalization.
 *
 * @param pubMedID validatePubMedID validates a Pub Med ID.
 * @returns true if validates, false otherwise
 */
export function validatePMID(pubMedID: string): boolean {
  const normalizedID = normalizePMID(pubMedID);
  return rePMID.test(normalizedID);
}

File: pmid.go
package metadatatools

import (
	"regexp"
	"strings"
)

// PMIDPattern matches a normalized PubMed identifier (digits only).
const PMIDPattern = `^[0-9]+$`

// rePMID is the compiled form of PMIDPattern.
var rePMID = regexp.MustCompile(PMIDPattern)

// rePMIDPrefix matches a leading "pmid:" label, with or without
// trailing whitespace.
var rePMIDPrefix = regexp.MustCompile(`(?i)pmid:\s*`)

// NormalizePMID removes a leading "pmid:" label and surrounding
// whitespace from pubMedID.
func NormalizePMID(pubMedID string) string {
	return strings.TrimSpace(rePMIDPrefix.ReplaceAllString(pubMedID, ""))
}

// ValidatePMID normalizes pubMedID then validates it against PMIDPattern.
func ValidatePMID(pubMedID string) bool {
	return rePMID.MatchString(NormalizePMID(pubMedID))
}
